xayn-ai/src/reranker/sync.rs

```diff
@@ -0,0 +1,157 @@
+use crate::{
+    data::{CoiPoint, UserInterests},
+    embedding::utils::l2_distance,
+    error::Error,
+    utils::nan_safe_f32_cmp_high,
+    CoiId,
+};
+use anyhow::bail;
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+const SOCIAL_DIST: f32 = 8.0;
+const CURRENT_SCHEMA_VERSION: u8 = 0;
+
+/// Synchronizable data of the reranker.
+#[cfg_attr(test, derive(Clone, PartialEq, Debug))]
+#[derive(Default, Serialize, Deserialize)]
+pub(crate) struct SyncData {
+    pub(crate) user_interests: UserInterests,
+}
+
+impl SyncData {
+    /// Deserializes a `SyncData` from `bytes`.
+    pub(crate) fn deserialize(bytes: &[u8]) -> Result<Self, Error> {
+        // version encoded in first byte
+        let version = bytes[0];
+        if version != CURRENT_SCHEMA_VERSION {
+            bail!(
+                "Unsupported serialized data. Found version {} expected {}.",
+                version,
+                CURRENT_SCHEMA_VERSION,
+            );
+        }
+
+        let data = bincode::deserialize(&bytes[1..])?;
+        Ok(data)
+    }
+
+    /// Serializes a `SyncData` to a byte representation.
+    pub(crate) fn serialize(&self) -> Result<Vec<u8>, Error> {
+        let size = bincode::serialized_size(self)? + 1;
+        let mut serialized = Vec::with_capacity(size as usize);
+        // version encoded in first byte
+        serialized.push(CURRENT_SCHEMA_VERSION);
+        bincode::serialize_into(&mut serialized, self)?;
+
+        Ok(serialized)
+    }
+
+    /// Synchronizes with another `SyncData`.
+    pub(crate) fn synchronize(&mut self, other: SyncData) {
```

Comment:
>i'm just worried that we have to clone the other at one point without need

well `SyncData` isn't `Clone` either - does that help assuage your worries?
otherwise happy to amend but `append_cois` will probably end up changing anyway so we can look at it in the follow up.

Reply:
ah i see, i overlooked that it implements `Clone` only for `cfg(test)`. sure makes sense to not spend too much effort if it's going to be changed anyways 👍